File: frontend/src/features/metrics/api/metric-definitions.ts
/**
 * Typed client for metric definition / catalog APIs (M4 + M5).
 * Paths use tenant UUID; store org_id on the server is the tenant slug.
 */

import { authedGet, authedRequest } from "@/lib/api";

export type MetricSource = "standard" | "standard_customized" | "custom" | "variant";

export type MetricCatalogItem = {
  metric_id: string;
  name: string;
  version: number;
  revision: number;
  status: string;
  content_hash: string | null;
  visibility: string | null;
  description: string | null;
  tags: string[];
  entity: string | null;
  source: MetricSource;
  extends: string | null;
  params_bound: Record<string, unknown> | null;
  draft_pending: boolean;
  notices: Array<{ id?: string; notice: string; payload?: Record<string, unknown> }>;
  compile_error: string | null;
  updated_at: string | null;
  enrolled: boolean;
};

export type MetricDefinitionDetail = {
  org_id: string;
  metric_id: string;
  version: number;
  revision: number;
  status: string;
  kind: string;
  yaml: string;
  resolved_json: Record<string, unknown> | null;
  content_hash: string | null;
  parent_pin: Record<string, unknown> | null;
  notices: Array<{ id?: string; notice: string; payload?: Record<string, unknown> }>;
  created_by: string | null;
  created_at: string | null;
};

export type MetricVersion = {
  metric_id: string;
  version: number;
  revision: number;
  status: string;
  content_hash: string | null;
  created_by: string | null;
  created_at: string | null;
  org_id: string;
};

export type MetricCatalogField = {
  name: string;
  type: string;
  role: string;
  values: string[] | null;
  nullable: boolean | null;
  cardinality_estimate: number | null;
  person: boolean;
  virtual: boolean;
  mapping_id: string | null;
};

export type MetricCatalogEntity = {
  name: string;
  grain: string | null;
  dbt_model: string | null;
  fields: MetricCatalogField[];
};

export type MetricCatalogResponse = {
  catalog_version: number;
  cardinality: Record<string, number>;
  entities: MetricCatalogEntity[];
  virtual_dimensions: Array<{
    mapping_id: string;
    entity: string;
    from_field: string;
    to_dimension: string;
  }>;
};

export type DiffResponse = {
  changes: Array<{
    path: string;
    op: string;
    before?: unknown;
    after?: unknown;
  }>;
  summary: string[];
  from_resolved: Record<string, unknown> | null;
  to_resolved: Record<string, unknown> | null;
};

export type MetricSetRead = {
  org: string;
  yaml: string | null;
  doc: Record<string, unknown>;
  version: number | null;
  status: string;
};

export type DimensionMappingSummary = {
  mapping_id: string;
  entity: string | null;
  from_field: string | null;
  to_dimension: string | null;
  version: number;
  status: string;
};

export type CompileRun = {
  id: string;
  started_at: string;
  finished_at: string | null;
  status: string;
  trigger: string;
  report_json: Record<string, unknown> | null;
};

export type MetricHealthSummary = {
  broken_count: number;
  notice_count: number;
  open_parent_version_notices: number;
  recent_compile_runs: CompileRun[];
  broken_metrics: Array<{
    metric_id: string;
    version: number;
    content_hash: string | null;
  }>;
};

export type GeneratedSql = {
  metric_id: string;
  content_hash: string | null;
  sql: string;
  source: "file" | "db" | "missing";
};

function base(tenantId: string) {
  return `/api/v1/tenants/${tenantId}`;
}

export function listMetricDefinitions(
  token: string,
  tenantId: string,
  options: {
    referencable?: boolean;
    entity?: string;
    includeDrafts?: boolean;
    includeBroken?: boolean;
  } = {},
): Promise<MetricCatalogItem[]> {
  const params = new URLSearchParams();
  if (options.referencable) params.set("referencable", "true");
  if (options.entity) params.set("entity", options.entity);
  if (options.includeDrafts === false) params.set("include_drafts", "false");
  if (options.includeBroken === false) params.set("include_broken", "false");
  const qs = params.toString();
  return authedGet(`${base(tenantId)}/metric-definitions${qs ? `?${qs}` : ""}`, token);
}

export function getMetricCatalog(
  token: string,
  tenantId: string,
): Promise<MetricCatalogResponse> {
  return authedGet(`${base(tenantId)}/metric-catalog`, token);
}

export function getMetricDefinition(
  token: string,
  tenantId: string,
  metricId: string,
): Promise<MetricDefinitionDetail> {
  const params = new URLSearchParams({ metric_id: metricId });
  return authedGet(`${base(tenantId)}/metric-definitions/detail?${params}`, token);
}

export function listMetricVersions(
  token: string,
  tenantId: string,
  metricId: string,
): Promise<MetricVersion[]> {
  const params = new URLSearchParams({ metric_id: metricId });
  return authedGet(`${base(tenantId)}/metric-definitions/versions?${params}`, token);
}

export function getMetricSql(
  token: string,
  tenantId: string,
  metricId: string,
): Promise<GeneratedSql> {
  const params = new URLSearchParams({ metric_id: metricId });
  return authedGet(`${base(tenantId)}/metric-definitions/sql?${params}`, token);
}

export function diffMetricDefinitions(
  token: string,
  tenantId: string,
  body: {
    metric_id: string;
    from_version?: number | null;
    to_version?: number | null;
    from_yaml?: string | null;
    to_yaml?: string | null;
  },
): Promise<DiffResponse> {
  return authedRequest(
    "POST",
    `${base(tenantId)}/metric-definitions:diff`,
    token,
    body,
  );
}

export function validateMetricDefinition(
  token: string,
  tenantId: string,
  yaml: string,
): Promise<{ ok: boolean; errors: unknown[]; warnings: unknown[] }> {
  return authedRequest("POST", `${base(tenantId)}/metric-definitions:validate`, token, {
    yaml,
  });
}

export function createMetricDefinition(
  token: string,
  tenantId: string,
  yaml: string,
): Promise<MetricDefinitionDetail> {
  return authedRequest("POST", `${base(tenantId)}/metric-definitions`, token, {
    yaml,
  });
}

export function putMetricDefinitionDraft(
  token: string,
  tenantId: string,
  body: {
    yaml: string;
    expected_version?: number | null;
    expected_revision?: number | null;
  },
): Promise<MetricDefinitionDetail> {
  return authedRequest(
    "PUT",
    `${base(tenantId)}/metric-definitions/draft`,
    token,
    body,
  );
}

export function classifyMetricDefinition(
  token: string,
  tenantId: string,
  body: { yaml: string; previous_version?: number | null },
): Promise<{
  kind: "none" | "non_semantic" | "semantic";
  next_version: number;
  next_revision: number;
  previous_version: number | null;
  previous_revision: number | null;
}> {
  return authedRequest(
    "POST",
    `${base(tenantId)}/metric-definitions:classify`,
    token,
    body,
  );
}

export function activateMetricDefinition(
  token: string,
  tenantId: string,
  metricId: string,
  body: { version?: number | null } = {},
): Promise<MetricDefinitionDetail> {
  const params = new URLSearchParams({ metric_id: metricId });
  return authedRequest(
    "POST",
    `${base(tenantId)}/metric-definitions:activate?${params}`,
    token,
    body,
  );
}

export function repinMetricDefinition(
  token: string,
  tenantId: string,
  metricId: string,
): Promise<MetricDefinitionDetail> {
  const params = new URLSearchParams({ metric_id: metricId });
  return authedRequest(
    "POST",
    `${base(tenantId)}/metric-definitions:repin?${params}`,
    token,
    {},
  );
}

export function archiveMetricDefinition(
  token: string,
  tenantId: string,
  metricId: string,
): Promise<MetricDefinitionDetail> {
  const params = new URLSearchParams({ metric_id: metricId });
  return authedRequest(
    "POST",
    `${base(tenantId)}/metric-definitions:archive?${params}`,
    token,
    {},
  );
}

export type PreviewResponse = {
  rows: Array<Record<string, unknown>>;
  timing_ms: number;
  sql: string;
  grain: string | null;
  diagnostics: Array<Record<string, unknown>>;
  truncated: boolean;
  sampled: boolean;
  executed: boolean;
  metric_id: string | null;
};

export function previewMetricDefinition(
  token: string,
  tenantId: string,
  yaml: string,
): Promise<PreviewResponse> {
  return authedRequest("POST", `${base(tenantId)}/metric-definitions:preview`, token, {
    yaml,
  });
}

export function getMetricSet(token: string, tenantId: string): Promise<MetricSetRead> {
  return authedGet(`${base(tenantId)}/metric-set`, token);
}

export function putMetricSet(
  token: string,
  tenantId: string,
  yaml: string,
): Promise<MetricSetRead> {
  return authedRequest("PUT", `${base(tenantId)}/metric-set`, token, { yaml });
}

export function listDimensionMappings(
  token: string,
  tenantId: string,
): Promise<DimensionMappingSummary[]> {
  return authedGet(`${base(tenantId)}/dimension-mappings`, token);
}

export function getMetricHealth(
  token: string,
  tenantId: string,
): Promise<MetricHealthSummary> {
  return authedGet(`${base(tenantId)}/metric-health`, token);
}

export function listCompileRuns(
  token: string,
  tenantId: string,
): Promise<CompileRun[]> {
  return authedGet(`${base(tenantId)}/metric-compile-runs`, token);
}

File: backend/app/routers/metric_definitions.py
"""Tenant-scoped metric definition management (M4).

Uses query param ``metric_id`` for dotted ids (e.g. propel.cycle_time).
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth.dependencies import require_permission
from app.auth.manager import current_active_user
from app.db.session import get_async_session
from app.models.user import User
from app.schemas.metric_definitions import (
    ActivateBody,
    CompileRunRead,
    MetricDefinitionRead,
    MetricSetRead,
    MetricSummaryRead,
    ValidateResponse,
    YamlBody,
)
from app.services import metric_definitions as svc

router = APIRouter(prefix="/api/v1", tags=["metric-definitions"])


@router.get(
    "/tenants/{tenant_id}/metric-definitions",
    response_model=list[MetricSummaryRead],
)
async def list_metric_definitions(
    ctx=Depends(require_permission("metrics:read")),
    session: AsyncSession = Depends(get_async_session),
):
    rows = await svc.list_resolved_summaries(session, ctx.tenant.slug)
    return [MetricSummaryRead.model_validate(r) for r in rows]


@router.get(
    "/tenants/{tenant_id}/metric-definitions/detail",
    response_model=MetricDefinitionRead,
)
async def get_metric_definition(
    metric_id: str = Query(..., description="Full namespaced metric id"),
    ctx=Depends(require_permission("metrics:read")),
    session: AsyncSession = Depends(get_async_session),
):
    detail = await svc.get_definition_detail(session, ctx.tenant.slug, metric_id)
    return MetricDefinitionRead.model_validate(detail)


@router.post(
    "/tenants/{tenant_id}/metric-definitions:validate",
    response_model=ValidateResponse,
)
async def validate_metric_definition(
    body: YamlBody,
    ctx=Depends(require_permission("metrics:read")),
):
    _ = ctx
    return ValidateResponse.model_validate(await svc.validate_definition(body.yaml))


@router.post(
    "/tenants/{tenant_id}/metric-definitions",
    response_model=MetricDefinitionRead,
    status_code=201,
)
async def create_metric_definition(
    body: YamlBody,
    ctx=Depends(require_permission("metrics:manage")),
    session: AsyncSession = Depends(get_async_session),
    user: User = Depends(current_active_user),
):
    row = await svc.create_draft(
        session,
        ctx.tenant.slug,
        body.yaml,
        created_by=str(user.id),
    )
    detail = await svc.get_definition_detail(session, ctx.tenant.slug, row.metric_id)
    return MetricDefinitionRead.model_validate(detail)


@router.post(
    "/tenants/{tenant_id}/metric-definitions:activate",
    response_model=MetricDefinitionRead,
)
async def activate_metric_definition(
    body: ActivateBody,
    metric_id: str = Query(...),
    ctx=Depends(require_permission("metrics:manage")),
    session: AsyncSession = Depends(get_async_session),
):
    row = await svc.activate_definition(
        session, ctx.tenant.slug, metric_id, version=body.version
    )
    detail = await svc.get_definition_detail(session, ctx.tenant.slug, row.metric_id)
    return MetricDefinitionRead.model_validate(detail)


@router.post(
    "/tenants/{tenant_id}/metric-definitions:repin",
    response_model=MetricDefinitionRead,
)
async def repin_metric_definition(
    metric_id: str = Query(...),
    ctx=Depends(require_permission("metrics:manage")),
    session: AsyncSession = Depends(get_async_session),
    user: User = Depends(current_active_user),
):
    row = await svc.repin_definition(
        session, ctx.tenant.slug, metric_id, created_by=str(user.id)
    )
    detail = await svc.get_definition_detail(session, ctx.tenant.slug, row.metric_id)
    return MetricDefinitionRead.model_validate(detail)


@router.post(
    "/tenants/{tenant_id}/metric-definitions:archive",
    response_model=MetricDefinitionRead,
)
async def archive_metric_definition(
    metric_id: str = Query(...),
    ctx=Depends(require_permission("metrics:manage")),
    session: AsyncSession = Depends(get_async_session),
):
    row = await svc.archive_definition(session, ctx.tenant.slug, metric_id)
    return MetricDefinitionRead(
        org_id=row.org_id,
        metric_id=row.metric_id,
        version=row.version,
        revision=row.revision,
        status=row.status,
        kind=row.kind,
        yaml=row.yaml,
        resolved_json=row.resolved_json,
        content_hash=row.content_hash,
        parent_pin=row.parent_pin,
        notices=[],
    )


@router.get(
    "/tenants/{tenant_id}/metric-set",
    response_model=MetricSetRead,
)
async def get_metric_set(
    ctx=Depends(require_permission("metrics:read")),
    session: AsyncSession = Depends(get_async_session),
):
    return MetricSetRead.model_validate(
        await svc.get_metric_set(session, ctx.tenant.slug)
    )


@router.put(
    "/tenants/{tenant_id}/metric-set",
    response_model=MetricSetRead,
)
async def put_metric_set(
    body: YamlBody,
    ctx=Depends(require_permission("metrics:manage")),
    session: AsyncSession = Depends(get_async_session),
    user: User = Depends(current_active_user),
):
    await svc.put_metric_set(
        session, ctx.tenant.slug, body.yaml, created_by=str(user.id)
    )
    return MetricSetRead.model_validate(
        await svc.get_metric_set(session, ctx.tenant.slug)
    )


@router.get("/tenants/{tenant_id}/dimension-mappings/detail")
async def get_dimension_mapping(
    mapping_id: str = Query(...),
    ctx=Depends(require_permission("metrics:read")),
    session: AsyncSession = Depends(get_async_session),
):
    return await svc.get_dimension_mapping(session, ctx.tenant.slug, mapping_id)


@router.put("/tenants/{tenant_id}/dimension-mappings")
async def put_dimension_mapping(
    body: YamlBody,
    ctx=Depends(require_permission("metrics:manage")),
    session: AsyncSession = Depends(get_async_session),
    user: User = Depends(current_active_user),
):
    row = await svc.put_dimension_mapping(
        session, ctx.tenant.slug, body.yaml, created_by=str(user.id)
    )
    return await svc.get_dimension_mapping(session, ctx.tenant.slug, row.metric_id)


@router.get(
    "/tenants/{tenant_id}/metric-compile-runs",
    response_model=list[CompileRunRead],
)
async def list_metric_compile_runs(
    ctx=Depends(require_permission("metrics:read")),
    session: AsyncSession = Depends(get_async_session),
):
    _ = ctx
    rows = await svc.list_compile_runs(session)
    return [CompileRunRead.model_validate(r) for r in rows]
